src/Arrays/Buffers.jl

```diff
@@ -0,0 +1,232 @@
+module Buffers
+
+using Requires
+using MPI
+using CUDAapi
+
+export Buffer
+export SingleBuffer, DoubleBuffer
+export get_stage, get_transfer, prepare_transfer!, prepare_stage!
+
+@enum BufferKind begin
+  SingleBuffer
+  DoubleBuffer
+end
+
+###
+# Note: We use pinned and device-mapped hostbuffers in double staging
+# Potential improvements
+# - Investigate if we need to device-map
+# - Implement single buffered with pinned + device-mapped hostbuffers
+
+"""
+    Buffer{T}(::Type{Arr}, kind, dims...; pinned = true)
+
+Optionally double buffered space for MPI communication.
+Call `get_stage()` to get a buffer into which to stage
+data for communication.
+Call `get_transfer()` to get the buffer
+to use for communication.
+
+In the case that double buffering is not needed, we avoid
+the extra copy and use the staging buffer for transfers.
+
+If we're running on the host only, double buffering is not
+necessary. If we're running on the GPU and we have
+CUDA-enabled MPI, we can avoid double buffering and use the
+GPU memory directly. If we do not have CUDA-enabled MPI, we
+must use double buffering, where the staging buffer is on
+the device and the transfer buffer is on the host.
+
+# Arguments
+- `T`: Element type
+- `Arr::Type`: What kind of array to allocate for `stage`
+- `kind::BufferKind`: Either `Single` or `Double`
+- `dims...`: Dimensions of the array
+
+# Keyword Arguments
+- `pinned`:  Register the `transfer` buffer with CUDA
+"""
+struct Buffer{T, Arr, Buff}
+  stage :: Arr     # Same type as Q.data, used for staging
+  transfer :: Buff # Union{Nothing,Buff}
+
+  function Buffer{T}(::Type{Arr}, kind, dims...; pinned = false) where {T, Arr}
+    if kind == SingleBuffer
+      transfer = nothing
+    elseif kind == DoubleBuffer
+      transfer = zeros(T, dims...)
+      if pinned
+        # XXX: make this
+        # transfer = register(transfer) to obtain a HostBuffer
+        # or try:
+        # ```
+        #   transfer = register(transfer)::HostBuffer
+        #   stage    = # CuArray(transfer)
+        # ```
+        register(transfer)
+        # XXX: Should this finalizer be attached to the Buffer struct?
+        finalizer(unregister, transfer)
+      end
+    else
+      error("Bufferkind $kind is not implemented yet")
+    end
+
```

Comment:
Why is pinnned memory turned off?  As far as I understand it, we need pinned memory to overlap communication and computation.

I am not sure it is correct but here
https://github.com/lcw/Heptapus.jl/blob/0f235861c201cbfdcfc4368629e1dccda3868b57/examples/mpicuda/try.jl#L6
is an example of how I registered pinned memory to achieve overlapping communication and computation.
￼

Reply:
>  As far as I understand it, we need pinned memory to overlap communication and computation.

IIUC that is not tied to pinned memory, but to using non-blocking streams + async memory transfers.

> Why is pinnned memory turned off?

I was trying to debug the freezes bors is experiencing.

Reply:
I see.  I always though I needed pinned memory for async memory transfers.

Reply:
FYI, see for example Overlapping Kernel Execution and Data Transfers section of <https://devblogs.nvidia.com/how-overlap-data-transfers-cuda-cc>.